backend/src/apis/fetch.rs

```diff
@@ -1,151 +1,102 @@
-        for (i, balance) in balances.iter().enumerate() {
-            if asset_sums.len() <= i {
-                asset_sums.push(*balance);
-            } else {
-                let sum = asset_sums[i] + balance;
-                asset_sums[i] = sum;
-            }
-        }
-    }
-
-    Ok(asset_sums)
+use futures::future::try_join_all;
+use std::{error::Error, sync::Arc};
+
+use ethers::{
+    abi::Address,
+    contract::builders::ContractCall,
+    prelude::SignerMiddleware,
+    providers::{Http, Middleware, Provider},
+    signers::LocalWallet,
+    types::{H160, U256},
+};
+
+use crate::contracts;
+use contracts::generated::mock_erc20::MockERC20;
+
+pub trait BalanceFromContract<M: Middleware> {
+    fn get_balance_from_contract(&self, account: Address) -> ContractCall<M, U256>;
 }
 
 /// This function takes a list of token contracts, an address and returns the balances of that address for the queried contracts.
-#[tokio::main]
-async fn fetch_balances_per_addr(
-    address: String,
-    asset_contract_addresses: Vec<String>,
-) -> Result<Vec<u64>, Box<dyn Error>> {
-    // Create a header map
-    let mut headers = HeaderMap::new();
-
-    // Load .env file
-    dotenv::dotenv().ok();
-
-    // Access API key from the environment
-    let api_key = std::env::var("COVALENT_API_KEY").unwrap();
-
-    // Add `Content-Type` header
-    headers.insert(CONTENT_TYPE, HeaderValue::from_static("application/json"));
-
-    // Add `Authorization` header
-    let auth_value = format!("Basic {}", encode(api_key));
-    headers.insert(AUTHORIZATION, HeaderValue::from_str(&auth_value)?);
-
-    // Form URL
-    let url = format!(
-        "https://api.covalenthq.com/v1/eth-mainnet/address/{}/balances_v2/",
-        address
-    );
-
-    // Send a GET request to the API
-    let res = reqwest::Client::new()
-        .get(&url)
-        .headers(headers)
-        .send()
-        .await?
-        .json::<Response>()
-        .await?;
-
-    // Get balances for the specific tokens
-    let filtered_balances =
-        filter_balances_by_token_contracts(asset_contract_addresses, &res).unwrap();
-
-    Ok(filtered_balances)
-}
-
-/// This function filters out only the balances corresponding to an asset listed in asset_contract_addresses.
-fn filter_balances_by_token_contracts(
-    asset_contract_addresses: Vec<String>,
-    response: &Response,
-) -> Result<Vec<u64>, &'static str> {
-    let mut balances = Vec::new();
-    for contract_adddress in asset_contract_addresses {
-        if let Some(item) = response.data.items.iter().find(|&item| {
-            item.contract_address.to_ascii_lowercase() == contract_adddress.to_ascii_lowercase()
-        }) {
-            match item.balance.parse::<u64>() {
-                Ok(num) => balances.push(num),
-                Err(e) => println!("Failed to parse string: {}", e),
-            }
-        } else {
-            balances.push(0 as u64);
+pub async fn fetch_asset_sums<'a, M: Middleware + 'a>(
```

Comment:
Can you explain what are the inputs of such function? 

The way it worked before was `address` is the wallet address we are querying the `asset_sums` and `asset_contract_addresses` were the ERC20 token contract addresses

Reply:
Sure! Extend previous comment https://github.com/summa-dev/summa-solvency/pull/116#discussion_r1279164154

we now have three inputs in the `fetch_asset_sum` function: `client`, `contracts`, and `asset_addresses`.

You can think of the `client` as a provider responsible for fetching ETH balances. Meanwhile, `contracts` is an array of `ContractInstance`, with each element representing a high-level interface connected to a specific contract address via the provider. This interface allows for easy interaction with the smart contract.

Both `client` and `contracts` fetch balances by iterating through `asset_addresses`.

I chose to use `contracts` as an input because it offers more flexibility to adapt to different types of ERC contracts, not just ERC20.

Reply:
Understood!